src/engine/Service.cpp

```diff
@@ -241,65 +242,84 @@ void Service::writeJsonResult(const std::vector<std::string>& vars,
 }
 
 // ____________________________________________________________________________
-Result::Generator Service::computeResultLazily(
-    const std::vector<std::string> vars,
-    ad_utility::LazyJsonParser::Generator body, bool singleIdTable) {
-  LocalVocab localVocab{};
-  IdTable idTable{getResultWidth(), getExecutionContext()->getAllocator()};
+Result::LazyResult Service::computeResultLazily(
+    std::vector<std::string> vars, ad_utility::LazyJsonParser::Generator body,
+    bool singleIdTable) {
+  auto inputRange = ad_utility::CachingTransformInputRange(
+      std::move(body), [](auto& input) { return std::move(input); });
+  auto get = [service = this, vars = std::move(vars),
+              singleIdTable = singleIdTable, inputRange = std::move(inputRange),
+              localVocab = LocalVocab{},
+              idTable = IdTable{getResultWidth(),
+                                getExecutionContext()->getAllocator()},
+              rowIdx = size_t{0}, varsChecked = false, resultExists = false,
+              singleIdTableReturned =
+                  false]() mutable -> std::optional<Result::IdTableVocabPair> {
+    if (singleIdTableReturned) {
+      return std::nullopt;
+    }
+    try {
+      auto partJsonOpt = inputRange.get();
+      while (partJsonOpt.has_value()) {
```

Comment:
you can simplify this by using
`while (auto partJsonOpt = inputRange.get())` 
(then you get rid of the two explicit get lines, especially the one at the end of the loop is error-prone